### ISSUE TITLE
RuleTester does not currently support nested test cases

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#7500]
This is an issue for #7430.

### Test cases can be unreadable when lumped together

When running Mocha against some of the core rules' unit test cases, the test names are just the code snippets used in each test case. Sometimes it's really hard to understand what else might be going on, such as any rule options that are being used in those cases.

Here's an example output for a fairly simple set of tests for [brace-style](http://eslint.org/docs/rules/brace-style): https://gist.github.com/platinumazure/0a635d51122c37e59397767b7aba4bc2

Even with this fairly simple output, some test cases go onto the next line, so that complicates the reading of what should otherwise be a simple list of test cases.

### Test cases do not mention rule options or other pertinent information

Also, it's not clear what options are in effect for each test, so one has to guess *why* a particular test is valid/invalid, especially in the case where a code snippet is valid with one set of options and invalid with another.

### Typing the same options is a pain

A related pain point is that it can be a pain to type out the same options for a bunch of test cases.

Annoying:

```js
ruleTester.run(rule, {
    valid: [
        // ...
        { code: "someCode();", options: [{ foo: "bar" }] },
        { code: "someOtherCode();", options: [{ foo: "bar" }] },
        { code: "yetMoreCode();", options: [{ foo: "bar" }] },
        { code: "sickOfTypingTheSameOptions();", options: [{ foo: "bar" }] },
        // ...
    ],
    invalid: [
        // ...
    ]
});
```

It would be awesome to be able to use `Array.prototype.map` or similar to allow consecutive test cases to have the same options. However, when working with a single array of test cases, it is very difficult to apply a `map()` to only some of the array members. It can be done with `.concat()` and other similar methods, but it's a pain.

## Proposed Solution

My proposed solution is to augment RuleTester to support nested test cases (which I've called "scenarios"), so that developers can choose to group tests a certain way.

This improves all three problems above:

1. Test output would now show extra levels of nesting when developer chooses to group tests into a scenario
1. If developer employs useful scenario names (i.e., specifies what options are used), output becomes much more readable because a test case's parents show the necessary information
1. Since tests are now broken down into a separate array per scenario, setting up common options via `.map()` becomes much more feasible.

## Why not rely on the test framework?

@nzakas had raised a great question in the PR, which was this: Why do anything here, why not rely on the test framework?

The answer is that RuleTester fully controls its interactions with any test framework and encapsulates them via `RuleTester.describe` and `RuleTester.it`, and so in order to access nested test functionality in Mocha or any other test framework, we have to call those methods. Mocha has no facility for inferring that the user wants a nested test case, so we need to tell Mocha somehow-- and in order to be framework-agnostic, we need to do so through our `RuleTester.describe` and `RuleTester.it` contract.

## Could this break custom integrations between RuleTester and other test frameworks?

Actually, maybe. For example, old versions of QUnit had no way of handling nested modules whatsoever. That doesn't mean those integrations couldn't be tweaked, but it was something I hadn't considered on drafting the PR. So maybe this is a breaking change after all. Unfortunately, we won't know unless we can test some of those integrations with these changes.

**Comment by platinumazure** [comment]
Adding "breaking" label for now. To be clear, this only could break people using a non-Mocha test framework where they had to define `RuleTester.describe` and `RuleTester.it`, and where they had to do so in a (shall we say) non-reentrant manner.

**Comment by nzakas** [comment]
Re: Using Mocha, why can't we just do this?

```js

describe("First bunch of tests", function() {

 ruleTester.run(rule, {
    valid: [
        // ...
        { code: "someCode();", options: [{ foo: "bar" }] },
        { code: "someOtherCode();", options: [{ foo: "bar" }] },
        { code: "yetMoreCode();", options: [{ foo: "bar" }] },
        { code: "sickOfTypingTheSameOptions();", options: [{ foo: "bar" }] },
        // ...
    ],
    invalid: [
        // ...
    ]
 });

});

describe("second bunch of tests", function() {
 ruleTester.run(rule, {
    valid: [
        // ...
        { code: "someCode();", options: [{ foo: "bar" }] },
        { code: "someOtherCode();", options: [{ foo: "bar" }] },
        { code: "yetMoreCode();", options: [{ foo: "bar" }] },
        { code: "sickOfTypingTheSameOptions();", options: [{ foo: "bar" }] },
        // ...
    ],
    invalid: [
        // ...
    ]
 });
});
```

My overall point is that I don't think `RuleTester` should start to encompass things that really are testing-framework features. It's designed to be a tool for use with testing frameworks, not to replace them.

If you feel like some description might be useful for some tests, we could also add a `title` or `desc` property to the object alongside `code` and output that with the code, I'm just not sure going hierarchical is right for `RuleTester`.

**Comment by platinumazure** [comment]
@nzakas We could certainly encourage that, but I could take that argument to the next step and argue, why invoke describe/it at all in RuleTester? Why not require users to write code like this:

```js
describe("my-rule", function() {
    it("a simple valid case", function() {
        ruleTester.runValidCase("someCode()", { options: [], parserOptions: {} });
    });

    it("a valid case with options", function() {
         ruleTester.runValidCase("someOtherCode()", { options: [{ foo: "bar" }] });
    });
});
```

Answer: Because that would be a royal pain in the ass for users to type.

What I'm trying to say is we've already conceded that purity of code (in favor of an encapsulated, overrideable implementation which can conform to certain test frameworks) is worth trading off for convenience in writing test cases. My proposal just takes that a step further with nested test cases.

Using a property like `title` or `desc` doesn't solve the issue of users having to type the same options repeatedly.

**Comment by platinumazure** [comment]
Another argument in favor of this proposal is the ability for a plugin maintainer to switch test frameworks without having to touch test files. If we support nested test cases in RuleTester and if a plugin maintainer is able to get `RuleTester.describe` and `RuleTester.it` to play nicely with a new test framework, then the test files themselves wouldn't have to change. Requiring users to use `describe`/`it` around test cases means the test files are no longer framework-agnostic.

**Comment by nzakas** [comment]
> why invoke describe/it at all in RuleTester?

We do that so that it can plug into a testing framework seamlessly. You don't have to use a testing framework with `RuleTester`, it will still work on its own. However, we wanted to provide hooks that made it easier to interact with testing frameworks. `RuleTester` originally required Mocha before we made this change to decouple from testing framework. Otherwise, it would be hard to get your test results to look pretty using test framework reporters. 

> Another argument in favor of this proposal is the ability for a plugin maintainer to switch test frameworks without having to touch test files. 

This isn't really a design goal for `RuleTester`. We let people overwrite `describe` and `it`, so switching test frameworks is basically two lines of code if the testing framework doesn't have those natively.

**Comment by ilyavolodin** [comment]
I agree with @nzakas that this is more than what RuleTester was designed for. On the other hand, this is small enough change in the off-the-beaten pass part of code that doesn't get changed all that often (and as such doesn't require a lot of maintenance) that could benefit small group of people. I'm honestly not sure if it's worth merging something like this in. I think I'm slightly leaning towards not needing this, as there are workaround for the cases described by @platinumazure 

@platinumazure Another way to achieve this behavior without creating direct dependency on test framework in your code would be to use what @nzakas suggested (wrap your RuleTester cases into `describe` and `it`), but do that by using `RuleTester.describe` instead. This will not require any changes to RuleTester and should work just fine, even if you change testing framework.

**Comment by platinumazure** [comment]
@nzakas @ilyavolodin I like what @ilyavolodin has proposed with wrapping cases in `RuleTester.describe` and `RuleTester.it`. I will try to get a test repo up in the next few days which will use the exact same rule testing code, and see how it goes with Mocha vs no framework vs another framework (possibly with a bridging layer if needed). If it seems fully interoperable as is, then I can close this issue.

**Comment by nzakas** [comment]
Can we close this now?
